File: ruby/21.rb
# time ruby 21.rb
# 31626
# 3.10user 0.40system 0:03.58elapsed 97%CPU (0avgtext+0avgdata 0maxresident)k
# 0inputs+0outputs (0major+1683minor)pagefaults 0swaps

def divisors(n)
  d = (2..Math.sqrt(n)).select { |q| n%q == 0 }
  d = d + d.map { |q| n / q } + [1]
  d.uniq
end

def sigma(n)
  divisors(n).inject(&:+)
end

limit = 10000
result = (2..limit).map { |n| [n, sigma(n)] }.
                    select { |t| t.first > t.last && t.first == sigma(t.last) }.
                    map { |t| t.first + t.last }.
                    inject(&:+)
puts result

File: haskell/21.hs
-- time runghc 21.hs
-- 31626
-- 1.24user 0.07system 0:01.41elapsed 93%CPU (0avgtext+0avgdata 0maxresident)k
-- 16inputs+0outputs (1major+7926minor)pagefaults 0swaps

import Data.List (nub)

divisors n = let sqn = floor . sqrt . fromIntegral $ n
                 a = filter (\q -> mod n q == 0) [2..sqn]
                 b = map (\q -> div n q) a
             in nub ([1] ++ a ++ b)

sigma n = sum . divisors $ n

range = [2..10000]
result = let tuples = zip range (map sigma range)
             amicables = filter (\(a,b) -> a > b && a == sigma b) tuples
         in sum $ map (\(a,b) -> a+b) amicables

main = print result
